### PR TITLE
Majors crud operations

### PR DESCRIPTION
**Pull request by rohanmaheshwari430**
### Summary 

This PR builds off of the firebase operations done in the "Initialized Firebase and Defined API Endpoints" PR. This time, I implemented the Majors schema and CRUD Operations for the Majors collection. I also created the majors collection in firebase and populated it with with a couple of test objects to ensure API functionality. I also reorganized the types defined in the server file to a new file to ensure clean code. All endpoints were tested via postman. The object structure and request parameters can be found in in the attached Postman collection. 

### Test Plan 

To test these endpoints, cd into the server directory and type the command "yarn start". It may return an error saying you are missing the service account file, in which case please reach out to Jessica to get instructions on receiving a service account file. Make sure to rename it to the service account file name that the error mentions. If no error was returned, copy this Postman link at https://www.getpostman.com/collections/c5ea29567bc40885aa52 and import it as a collection in your personal Postman workspace. Now, you should be all set to test the endpoints. Each of these requests targets one of the four Majors-related API endpoints. Feel free to change the request body or query parameters to test different data.

https://www.getpostman.com/collections/c5ea29567bc40885aa52

GET Majors
Screen Shot 2021-04-09 at 12 57 19 PM

DELETE Major
Screen Shot 2021-04-09 at 12 57 48 PM

CREATE Major
Screen Shot 2021-04-09 at 12 58 15 PM

UPDATE Major
Screen Shot 2021-04-09 at 12 58 38 PM

### Notes 

- It may be worth creating separate files for each collection's API endpoints.

### CODE REVIEW

**Review by zachary-kent — COMMENTED** [review]
Not positive about this, but I'm pretty sure most checks for example "field == null" should be "field === undefined" I'm pretty sure that if a field is not present in an object, it will be undefined rather than null. Also, you should use === as opposed to == for stricter equality checking and to avoid undesired type coercion.

**Comment by rohanmaheshwari430** [comment]
> Not positive about this, but I'm pretty sure most checks for example "field == null" should be "field === undefined" I'm pretty sure that if a field is not present in an object, it will be undefined rather than null. Also, you should use === as opposed to == for stricter equality checking and to avoid undesired type coercion.

Thanks for the critic! So I went back and tested my endpoint by purposely not including one of the fields, and I did end up getting a failure response saying that one or more fields was missing. So it seems like 'field == null' does the trick? I think this may be because "undefined" still means there is some value for that field, it is just not comprehendable. Null means there is no value for that field. What do you think?

**Comment by meganyin13** [comment]
@rohanmaheshwari430 great work on this PR! But I will back up Zak's point that you should use `===` _always_. This is just one of the ugly parts of JavaScript/TypeScript that ES6 tries to fix by adding `===`. Triple equals does strict equality checking whereas double equals is just checking if something is "truthy" or "falsy". This means JS will try to coerce whatever you're comparing to a true/false value and that means things like 0 or "" (empty string) can end up as false. Currently you're seeing `== undefined` works fine because it is falsy but it's better to strictify and explicitly say you want it to be `=== undefined` since things like `0 == undefined` or `"" == undefined` due to the falsy comparison. 
You can read more about the bad parts of JS here https://dev.to/khalalw/javascript-quirks-explaining-the-bad-parts-299j and this is the video everyone refers to when saying JS kinda sucks haha https://www.destroyallsoftware.com/talks/wat

Still great work on this PR and I'm happy to see such careful reviews!😁

**Comment by rohanmaheshwari430** [comment]
> @rohanmaheshwari430 great work on this PR! But I will back up Zak's point that you should use `===` _always_. This is just one of the ugly parts of JavaScript/TypeScript that ES6 tries to fix by adding `===`. Triple equals does strict equality checking whereas double equals is just checking if something is "truthy" or "falsy". This means JS will try to coerce whatever you're comparing to a true/false value and that means things like 0 or "" (empty string) can end up as false. Currently you're seeing `== undefined` works fine because it is falsy but it's better to strictify and explicitly say you want it to be `=== undefined` since things like `0 == undefined` or `"" == undefined` due to the falsy comparison.
> You can read more about the bad parts of JS here https://dev.to/khalalw/javascript-quirks-explaining-the-bad-parts-299j and this is the video everyone refers to when saying JS kinda sucks haha https://www.destroyallsoftware.com/talks/wat
> 
> Still great work on this PR and I'm happy to see such careful reviews!😁

Thank you for this advice @meganyin13 . Sorry for the misunderstanding (@zachary0kent). I will make the changes accordingly to enforce this stricter typing.

**Review by zachary-kent — APPROVED** [review]
Thanks for changing the '==' to '==='! This PR looks awesome and all of your endpoints seem to be implemented perfectly--I used most of them while writing code for other branches I was working on and ran into no issues. I also appreciate that you changed all of the snake case to camel case--I know that must've been tedious. My only suggestion is that you should probably avoid indexed for loops, replacing them with functional calls to map() and filter(), though it's probably not worth it to change it if it works now. Also, it might be helpful to document the type of the object/array each endpoint sends back. Once again, great job!

File: server/server.ts [hunk 1]
```diff
@@ -133,6 +161,300 @@ app.post('/updateCourse', async (req, res) => {
...
+app.post('/updateMajor', async (req, res) => {
+  const { field } = req.body;
+  const { title } = req.body;
+  const { content } = req.body;
+
+  if (content === null || field === null || title === null) {
+    res.send('One or more fields is missing.');
+  } else {
+    majors.doc(title).update({ field: content });
+    res.send({ success: true, data: [] });
```
Comment (jessfeng) [inline server/server.ts]: If data is just going to be an empty array, it might be better to just not include it when sending a response back.

**Review by jessfeng — COMMENTED** [review]
Looks great! Just had one comment about the response for the update endpoint, but other than that, we should be good to go!

File: server/server.ts (hunk 1 above)
Comment (rohanmaheshwari430) [reply server/server.ts]: Thanks for the comment, I agree with that change and have implemented it!

**Pull request merged**